### PR TITLE
[New] Generate data recording segment and function coverage over time.

### PR DESCRIPTION
**Pull request by BharathMonash**
Sorry about the new PR. Reviewing would be easier this way :)

PR is about generating 3 compressed CSV files recording the segment and function coverage over time. The CSV files (headers mentioned alongside) being generated are:

 ```
         1. segments.csv (only covered segments) - (benchmark_id, fuzzer_id, trial_id, time_stamp, file_id, line, col)
         2. functions.csv - (benchmark_id, fuzzer_id, trial_id, time_stamp, function_id, hits)
         3. names.csv  - (id, name, type)
```

The segment and function coverage is being recorded while measuring snapshots (every 900 secs), hence the time_stamp column. The generated data captures the concrete code elements that are covered over time. For functions, we also maintain hit counts. For segments, we only record those that have been covered and add the newly covered segments for later time stamps. This makes it space efficient alongside compression.

This data is generated over the entire campaign at every snapshot point when coverage is measured but the CSV files are only available after the experiment ends.

Filestore path for these files:  **`experiment_data/$(EXPERIMENT_NAME)/coverage/data/{files above}.csv`** 
Since the files contain information specific to an experiment for all benchmark, fuzzer, trial combinations I chose the path above as the destination to copy these files.

This data is being generated for issue #686
@mboehme : "We are interested in (unique) coverage over time. We would use this data for empirical studies of the utility of various coverage criteria as measures of fuzzer performance and to study how we can visualise the progress of a fuzzer's coverage frontier in a subject over time, or visualise fuzzer roadblocks (potentially viz-a-viz other fuzzers)."

```
Experiment config:
        benchmark: libjpeg-turbo-07-2017
        fuzzers: libfuzzer, honggfuzz
        trials: 3 (each)
        max_total_time: 1860 (2 cycles)

Output files size:
        segment.csv.gz:  104.1 kB
        functions.csv.gz: 22.9 kB
        names.csv.gz:  5.3 kB
```

Based on the experiment results shown above, if we have to estimate the sizes on a full experiment, it can be calculated as:
```((size_of_segment.csv.gz + (size_of_function.csv.gz * measure_cycles * trials)) * fuzzers + size_of_name.csv.gz) * benchmarks ```

So for a 24-hour experiment (96 cycles), 20 benchmarks and 15 fuzzers and 10 trials the estimated size is ~ 552 mB. Due to certain repetitive patterns in the data, the final compressed sized may even result to lower than 552 mB as compression engines give a better compression rate if all the records are similar in a way.

### CODE REVIEW

File: experiment/measurer.py [hunk 1]
```diff
@@ -89,15 +97,30 @@ def measure_loop(experiment: str, max_total_time: int):
...
                 all_trials_ended = scheduler.all_trials_ended(experiment)
 
-                if not measure_all_trials(experiment, max_total_time, pool, q):
+                if not measure_all_trials(experiment, max_total_time, pool, q,
+                                          segment_list, function_list):
                     # We didn't measure any trials.
                     if all_trials_ended:
+                        # Concatenate all data frames in the segment and
+                        # function multiprocessing list.
+                        experiment_specific_df_container.segment_df = (
```
Comment (BharathMonash) [inline experiment/measurer.py:119]: The trade-off here is that, this might be bit intensive in terms of memory, as the size data frame grows over time and is larger for longer experiments.

Printing ```segment_df.memory_usage(index=True, deep=True).sum()``` on the logger is in one way you could keep track of overall memory consumption.

Comment (mboehme) [reply experiment/measurer.py:119]: @BharathMonash Can you check the memory consumption for our local setup? The dispatcher runs on an n1-highmem-96 GCP instance, which has 624GB of main memory. If memory is a problem, we can merge and prune after each iteration of the measurer loop when the async call returns. Or we can write trial-specific CSVs to disk throughout, and load, merge and prune at the end.

Comment (BharathMonash) [reply experiment/measurer.py:119]: The memory consumption for the experiment config mentioned above in the PR description with the way that we are currently generating data is about 3.84 mb. With same, for a 24 hour experiment with 20 benchmarks,15 fuzzers and 10 trials the estimated memory consumption is ~ 8.23 GB. Pruning after each measure cycle reduces the estimate on a full experiment to 7.99 GB.

**Comment by inferno-chromium** [comment]
@jonathanmetzman has some other ideas to implement this, i think https://github.com/google/fuzzbench/pull/838. leaving to @jonathanmetzman for review here.

**Comment by mboehme** [comment]
Thanks @inferno-chromium & @jonathanmetzman! Love this idea. Sampling generated inputs live during fuzzing on the Runner would definitely give a better insight into where the fuzzer, *on the average*, spends most of its energy.

However, in this PR we are more interested in the corner cases, i.e., coverage elements that are *really* difficult to reach, like once in 100M generated test inputs (a 23h campaign that generates 10k inputs per second generates about 0.8 billion test inputs). The sampling-based approach can't have such a high sampling rate. Otherwise, it would be impractical.

We can use the data produced in this PR to investigate differential coverage across two or more fuzzers. We are planning to lift data on the probability of an element to be covered by looking across several trials (an element which appears in only 1 of 20 trials of 23 hours is less likely to be hit in random trial of 23hours than an element that appears in 20 of 20 trials). We already removed hit counts from segments.csv, and we can also remove hit counts from functions.csv.

As for reducing the memory consumption, @BharathMonash is working on the following: For each {benchmark, fuzzer, function, file} name, we compute a small truncated MD5 hash which is written to a multiprocessing dict. This way the process-specific segments and functions data frames can use these hashes without any concurrent reads, but still with a lower memory footprint. In each loop of the measurer, the segments data frame is merged such that segments covered in earlier trials are not added to the merged segments data frame.

**Comment by jonathanmetzman** [comment]
What is "[New]" supposed to convey in this description?

**Comment by jonathanmetzman** [comment]
I've taken a first pass on this PR.
Before I finish, I want to make sure of something.
Since this is a pretty invasive change can you commit to maintaining this code you are adding?
I wonder if there isn't a way to do this that is less invasive. Maybe postprocessing of corpora after measurement.

**Comment by mboehme** [comment]
@googlebot I consent

**Comment by mboehme** [comment]
> Since this is a pretty invasive change can you commit to maintaining this code you are adding?

Yes, my group will commit to maintaining this code.

> I wonder if there isn't a way to do this that is less invasive. Maybe postprocessing of corpora after measurement.

We actually tried a few things. The least invasive would be to store this data in {benchmark,fuzzer,trial,time}-specific folders, but without a regular pruning of redundant data (which we do in each measurer iteration @ `remove_redundant_duplicates`), this might run up to a few hundred GB of disk space. These are @BharathMonash's current values:
```
Experiment config:
Benchmark: Libjpeg-turbo-07-2017
fuzzers: libfuzzer, honggfuzz
max_total_time: 1860
trials: 3

Memory Consumption:
segment_df = 2.032 mb
function_df = 0.16 mb
name_df = 0.0195 mb

File size:
segments.csv.gz = 114.5 kb
functions.csv.gz = 16.4 kb
names.csv.gz = 7.5 kb
```
Should we run larger experiments to see memory consumption? What would be a reasonable configuration to test?

File: experiment/coverage_utils.py [hunk 2]
```diff
@@ -34,8 +36,8 @@
 
 
 def get_coverage_info_dir():
-    """Returns the directory to store coverage information including
-    coverage report and json summary file."""
+    """Returns the directory to store coverage information including coverage
```
[the remark this answers is not in the document]
Comment (BharathMonash) [reply experiment/coverage_utils.py]: Fixed.

File: experiment/test_measurer.py [hunk 3]
```diff
@@ -166,47 +166,56 @@ def test_generate_summary(mocked_get_coverage_binary, mocked_execute,
 @mock.patch('common.logs.error')
 @mock.patch('experiment.measurer.initialize_logs')
 @mock.patch('multiprocessing.Queue')
+@mock.patch('multiprocessing.list')
+@mock.patch('multiprocessing.list')
```
[the remark this answers is not in the document]
Comment (BharathMonash) [reply experiment/test_measurer.py]: Fixed.

File: experiment/test_measurer.py [hunk 4]
```diff
@@ -166,47 +166,56 @@ def test_generate_summary(mocked_get_coverage_binary, mocked_execute,
...
+    measurer.measure_trial_coverage(measure_request, max_cycle, mocked_queue(),
+                                    mocked_seg_list(), mocked_func_list())
     expected_calls = [
         mock.call(FUZZER, BENCHMARK, TRIAL_NUM, cycle)
         for cycle in range(min_cycle, max_cycle + 1)
     ]
     assert mocked_measure_snapshot_coverage.call_args_list == expected_calls
 
 
+@mock.patch('multiprocessing.list')
```
[the remark this answers is not in the document]
Comment (BharathMonash) [reply experiment/test_measurer.py]: Fixed.

File: experiment/test_measurer.py [hunk 5]
```diff
@@ -431,12 +441,24 @@ def test_measure_loop_end(_, __, ___, ____, _____, ______, experiment_config,
                           db_experiment):
     """Tests that measure_loop stops when there is nothing left to measure. In
     this test, there is nothing left to measure on the first call."""
-    measurer.measure_loop(experiment_config, 100)
+
+    # Register timeout handler and set timeout.
```
[the remark this answers is not in the document]
Comment (mboehme) [reply experiment/test_measurer.py:447]: The measure loop catches all exceptions *inside* the loop. During development, this test (to check whether the measurer is always terminating) froze because of a handled exception. The signal is simply a timeout to handle the freezing loop. We can add a comment or remove the code.

File: experiment/measurer.py [hunk 6]
```diff
@@ -376,7 +402,7 @@ def run_cov_new_units(self):
         self.UNIT_BLACKLIST[self.benchmark] = (
             self.UNIT_BLACKLIST[self.benchmark].union(set(crashing_units)))
 
-    def generate_summary(self, cycle: int, summary_only=True):
+    def generate_summary(self, cycle: int, summary_only=False):
```
[the remark this answers is not in the document]
Comment (BharathMonash) [reply experiment/measurer.py:411]: For recording function and segment coverage (to the detail that this PR is recording), we need the llvm to produce a full fledged coverage summary. With `summary_only=True` the coverage summary will have only high level information like {count, covered, percentage} for functions, instantiations and regions. But we need much more detailed information like segment line, column, filename, function hits and so on.

[the remark this answers is not in the document]
Comment (inferno-chromium) [reply experiment/measurer.py:411]: I definitely remember this part which is measurer will current fail if we do summary=True , it needs lots of ram and those 48-96 threads just break in parallel. this is the reason why we do it sequentially at end when summary=True. I dont have a great solution for this memory issue until we break measurer into individual bots (will probably come end of year after bug based benchmarking).

File: experiment/test_measurer.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (BharathMonash) [reply experiment/test_measurer.py:447]: @mboehme, I removed it already. Just making few more changes before I commit.

File: experiment/test_coverage_utils.py [hunk 7]
```diff
@@ -24,6 +24,83 @@ def get_test_data_path(*subpaths):
     return os.path.join(TEST_DATA_PATH, *subpaths)
 
 
+def test_extract_segments_and_functions_from_summary_json(fs):
+    """Tests that extract_covered_regions_from_summary_json returns the covered
+    segments and functions from summary json file."""
+    num_functions_in_cov_summary = 3  # for testing
```
[the remark this answers is not in the document]
Comment (BharathMonash) [reply experiment/test_coverage_utils.py:30]: Fixed.

File: experiment/measurer.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (mboehme) [reply experiment/measurer.py:411]: By default, [llvm-cov auto-detects an an appropriate number of threads to use](https://llvm.org/docs/CommandGuide/llvm-cov.html). Does `-num-threads=1` help?

[the remark this answers is not in the document]
Comment (inferno-chromium) [reply experiment/measurer.py:411]: So that one i did explore before, it is only for html report generation i remember
"Use N threads to write file reports (only applicable when -output-dir is specified)"
Maybe Bharath can verify this with local experiment, start 1 fuzzer against all benchmark and one trial and see measurer top memory usage. 

I am really sorry about these measurer limitations, we want to break it, but we have to do these one by one due to context switching with other things like ClusterFuzz,OSS-Fuzz.

[the remark this answers is not in the document]
Comment (mboehme) [reply experiment/measurer.py:411]: No worries, at all. If there is a limitation that we cannot resolve at our end, we'll give up on this PR (and wait for the live coverage stats). First, we'll see if there is an easy workaround, though. We'll come back on the memory consumption with one thread, one trial, one fuzzer, all benchmarks, for two snapshot iterations (like, 1860 seconds).

[the remark this answers is not in the document]
Comment (BharathMonash) [reply experiment/measurer.py:411]: Hi @inferno-chromium, I tried the the configuration that you suggested but against three benchmarks since I'm running fuzzbench locally. The configuration is mentioned below for your reference.

```
Experiment config (local):
        benchmark:  libjpeg-turbo-07-2017,  freetype2-2017,  harfbuzz-1.3.2
        fuzzers:  libfuzzer
        trials:  1
        max_total_time:  1860 (2 cycles)
```

So, with the `-summary-only` flag for the `llvm-cov export` call, the memory consumption by these three threads totalled to 2.43 GB.

On the flip side, without the `-summary-only` flag and with `-num-threads=1`, `-skip-expansions` and `-region-coverage-gt=0` flags the memory consumption totalled to 3.35 GB for these three multiprocess threads. 
Its safe to say that, the `-num-thread=1` option works without an output directory specified. `-skip-expansions` and  `-region-coverage-gt=0` would also help reduce the load on the `llvm-cov export` call when exporting full summary as it trims down all the parts of the JSON output that we do not want. Hope this helps :)

[the remark this answers is not in the document]
Comment (inferno-chromium) [reply experiment/measurer.py:411]: -num-threads=1, -skip-expansions are ok, but i think we cant use -region-coverage-gt=0 since we need info on what is not getting covered ? can you also try this on php benchmark, that one i remember was the most memory intensive.
Also, this code is getting heavily refactored by @jonathanmetzman to break up measurer on multiple bot, so i suggest waiting till that work is finished.

[the remark this answers is not in the document]
Comment (BharathMonash) [reply experiment/measurer.py:411]: No worries at all, I'll try this on php benchmark and get back to you on the memory consumption.

With `-region-coverage-gt=0`, I took the liberty of adding it, as `covered_regions.json` records information on regions with hit_count greater than 0. If it cannot be used, I'll remove it, no worries. https://github.com/google/fuzzbench/blob/b9853d50eb96fcd233f2a233457056e4122234fa/experiment/measurer/coverage_utils.py#L289.

Eagerly waiting on the queue based measurer :)

[the remark this answers is not in the document]
Comment (BharathMonash) [reply experiment/measurer.py:411]: On trying the same on `php_php-fuzz-parser` benchmark (only benchmark for the experiment), 
- The memory consumption by the `llvm-cov export` call with the `-summary-only` flag is about **214 MB** .
- llvm-cov export` call without the `-summary-only` flag and with `-num-threads=1`, `-skip-expansions` and `-region-coverage-gt=0` flags the memory consumption is **627 MB**.
